`ak.packed` casts categorical to value type

### Version of Awkward Array

1.10.0.rc2

### Description and code to reproduce

`ak.packed` seems to get rid of categoricals

```python
import awkward._v2 as ak

a = ak.Array(["foo", "foo", "bar"])
a_cat = ak.to_categorical(a)

assert a_cat.typestr == ak.packed(a_cat).typestr, f"'{a_cat.typestr}' != '{ak.packed(a_cat).typestr}'"
```

```pytb
      3 a = ak.Array(["foo", "foo", "bar"])
      4 a_cat = ak.to_categorical(a)
----> 6 assert a_cat.typestr == ak.packed(a_cat).typestr, f"'{a_cat.typestr}' != '{ak.packed(a_cat).typestr}'"

AssertionError: '3 * categorical[type=string]' != '3 * string'
```

This makes sense in that the indexed array is projected, but I did not expect `ak.packed` to change the type of an array.

The docstring also says this:

> Returns an array with the same type and values as the input, but with packed inner structures

--- a/src/awkward/_v2/contents/indexedarray.py
+++ b/src/awkward/_v2/contents/indexedarray.py
@@ -1244,7 +1244,16 @@ class IndexedArray(Content):
             raise ak._v2._util.error(AssertionError(result))
 
     def packed(self):
-        return self.project().packed()
+        if self.parameter("__array__") == "categorical":
+            return IndexedArray(
+                self._index,
+                self._content.packed(),
+                identifier=self._identifier,
+                parameters=self._parameters,
+                nplike=self._nplike,
+            )
+        else:
+            return self.project().packed()
 
     def _to_list(self, behavior, json_conversions):
         out = self._to_list_custom(behavior, json_conversions)
@@ -1261,8 +1270,8 @@ class IndexedArray(Content):
         return IndexedArray(
             index,
             content,
-            identifier=self.identifier,
-            parameters=self.parameters,
+            identifier=self._identifier,
+            parameters=self._parameters,
             nplike=nplike,
         )
 

--- a/src/awkward/_v2/contents/indexedoptionarray.py
+++ b/src/awkward/_v2/contents/indexedoptionarray.py
@@ -1684,10 +1684,20 @@ class IndexedOptionArray(Content):
 
     def packed(self):
         original_index = self._index.raw(self._nplike)
-
         is_none = original_index < 0
         num_none = self._nplike.index_nplike.count_nonzero(is_none)
-        if self._content.length > len(original_index) - num_none:
+        if self.parameter("__array__") == "categorical" or self._content.length <= (
+            len(original_index) - num_none
+        ):
+            return ak._v2.contents.IndexedOptionArray(
+                self._index,
+                self._content.packed(),
+                self._identifier,
+                self._parameters,
+                self._nplike,
+            )
+
+        else:
             new_index = self._nplike.index_nplike.empty(
                 len(original_index), dtype=original_index.dtype
             )
@@ -1699,15 +1709,6 @@ class IndexedOptionArray(Content):
             return ak._v2.contents.IndexedOptionArray(
                 ak._v2.index.Index(new_index, nplike=self.nplike),
                 self.project().packed(),
-                self._identifier,
-                self._parameters,
-                self._nplike,
-            )
-
-        else:
-            return ak._v2.contents.IndexedOptionArray(
-                self._index,
-                self._content.packed(),
                 self._identifier,
                 self._parameters,
                 self._nplike,

--- a/tests/v2/test_1688-pack-categorical.py
+++ b/tests/v2/test_1688-pack-categorical.py
@@ -9,9 +9,16 @@ numpy = ak.nplike.Numpy.instance()
 
 def test():
     this = ak._v2.to_categorical(["one", "two", "one", "three", "one", "four"])
+    assert ak._v2.is_categorical(this)
     # Ensure packing by itself doesn't change the type
-    assert ak._v2.packed(this).type == this.type
+    this_packed = ak._v2.packed(this)
+    assert this_packed.type == this.type
     # Ensure the categories match between the two
-    assert ak._v2.all(ak._v2.categories(ak._v2.packed(this)) == ak._v2.categories(this))
+    assert ak._v2.all(ak._v2.categories(this_packed) == ak._v2.categories(this))
+
     # Ensure the inner types match (ignoring the length change)
-    assert ak._v2.packed(this[:-1]).type.content == this.type.content
+    this_subset_packed = ak._v2.packed(this[:-1])
+    assert ak._v2.is_categorical(this_subset_packed)
+    assert this_subset_packed.type.content == this.type.content
+    # Ensure the categories match between the two
+    assert ak._v2.all(ak._v2.categories(this_subset_packed) == ak._v2.categories(this))

--- a/tests/v2/test_1688-pack-categorical.py
+++ b/tests/v2/test_1688-pack-categorical.py
@@ -1,0 +1,17 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+import pytest  # noqa: F401
+import awkward as ak  # noqa: F401
+
+
+numpy = ak.nplike.Numpy.instance()
+
+
+def test():
+    this = ak._v2.to_categorical(["one", "two", "one", "three", "one", "four"])
+    # Ensure packing by itself doesn't change the type
+    assert ak._v2.packed(this).type == this.type
+    # Ensure the categories match between the two
+    assert ak._v2.all(ak._v2.categories(ak._v2.packed(this)) == ak._v2.categories(this))
+    # Ensure the inner types match (ignoring the length change)
+    assert ak._v2.packed(this[:-1]).type.content == this.type.content

